src/fs/bootstrap_record.hh

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#pragma once
+
+#include "cluster.hh"
+#include "inode.hh"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/fs/block_device.hh>
+
+namespace seastar::fs {
+
+class invalid_bootstrap_record : public std::runtime_error {
+public:
+    invalid_bootstrap_record(const char* msg) : std::runtime_error(msg) {}
+};
+
+/// In-memory version of the record describing characteristics of the file system (~superblock).
+class bootstrap_record {
+public:
+    static constexpr uint64_t magic_number = 0x5343594c4c414653; // SCYLLAFS
+    static constexpr uint32_t max_shards_nb = 500;
+
+    struct shard_info {
+        cluster_id_t metadata_cl; /// cluster id of the first metadata log cluster
+        cluster_range range_cl; /// range of clusters for data for this shard
+
+        bool operator==(const shard_info &) const noexcept;
+        bool operator!=(const shard_info &rhs) const noexcept { return !(*this == rhs); }
+    };
+
+    uint64_t version; /// file system version
+    unit_size_t sector_size; /// sector size in bytes
+    unit_size_t cluster_size; /// cluster size in bytes
+    inode_t root_directory; /// root inode number
+    uint32_t shards_nb; /// number of file system shards
+    std::vector<shard_info> shards_info; /// basic informations about each file system shard
+
+    bootstrap_record() = default;
+    bootstrap_record(uint64_t version, unit_size_t sector_size, unit_size_t cluster_size,
+            inode_t root_directory, uint32_t shards_nb, std::vector<shard_info> shards_info)
```

Comment:
I think shards_nb as a constructor parameter is redundant because it can be inferred from shards_info.size().

Reply:
What do you think about removing shards_nb from bootstrap_record class?

Reply:
That is a good idea!

Reply:
And maybe add a public method that returns number of shards to make it more obvious how to get this value.